### PR TITLE
Andybranch

### PR DESCRIPTION
**Pull request by andykuang1**
Added multithreading functionality and edited code to allow for proper testing to make sure threading is actually being used. Created a new integration test in python to test the threading. This is done by issuing a partial request and then seeing that the requests are still filled out properly.

### CODE REVIEW

File: src/connection.cpp [hunk 1]
```diff
@@ -123,17 +122,76 @@ RequestHandler* Connection::chooseHandler() {
...
+    
+        handler_status = chooseHandler()->HandleRequest(*request_, response_);
+        // If request handling failed, invoke NotFound handler
+        // (In the future, we could invoke different handlers based on the type of
+        // error that occured, but for now just send everything bad to 404!) 
+        if (handler_status != RequestHandler::OK) {
+          std::cerr << "Error: handler returned status code " << handler_status << ".\n";
+          std::cerr << "Invoking default handler.\n";
+          default_handler_->HandleRequest(*request_, response_);
+          }
```
Comment (mckennagalvin) [inline src/connection.cpp:164]: align closing bracket with "if"

File: src/server.cpp [hunk 2]
```diff
@@ -153,7 +155,16 @@ void Server::run() {
   // have finished. While the server is running, there is always at least one
   // asynchronous operation outstanding: the asynchronous accept call waiting
   // for new incoming connections.
-  io_service_.run();
+  std::vector<boost::shared_ptr<boost::thread>> threadArray;
+
+  for (int i = 0; i < numThreads_; i++){
```
Comment (mckennagalvin) [inline src/server.cpp:160]: I think we should call this variable `num_threads_` for consistency

**Review by mckennagalvin — COMMENTED** [review]
👍

**Comment by mattdrag** [comment]
It looks good, I'm going to go ahead and merge.

**Pull request merged**